### PR TITLE
nixos/bcg: init module, nixos/mqtt2influxdb: init module

### PR DESCRIPTION
**Pull request by Cynerd**
###### Description of changes

Add Bigclown (Hardwario) gateway and companion software.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/mqtt2influxdb/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    influxdb
+    jsonpath-ng
+    paho-mqtt
+    py-expression-eval
+    pyaml
+    pycron
+    schema
+  ];
+
+  src = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mqtt2influxdb/default.nix:26]: src should be put after version

File: pkgs/development/python-modules/mqtt2influxdb/default.nix [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    rev = "v${version}";
+    sha256 = "YDgMoxnH4vCCa7b857U6iVBhYLxk8ZjytGziryn24bg=";
+  };
+
+  meta = with lib; {
+    homepage = "https://github.com/hardwario/bch-mqtt2influxdb";
+    description = "Flexible MQTT to InfluxDB Bridge";
+    platforms = platforms.linux;
+    license = licenses.mit;
+    maintainers = with maintainers; [cynerd];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/mqtt2influxdb/default.nix:38]:
```suggestion
    maintainers = with maintainers; [ cynerd ];
```

File: pkgs/development/python-modules/bch/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  version = "1.2.1";
+
+  propagatedBuildInputs = [
+    click
+    click-log
+    paho-mqtt
+    pyaml
+  ];
+
+  src = fetchFromGitHub {
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bch/default.nix:21]: same here

File: pkgs/development/python-modules/bcg/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    sed -ri 's/@@VERSION@@/${version}/g' \
+      bcg/__init__.py setup.py
+  '';
+
+  meta = with lib; {
+    homepage = "https://github.com/hardwario/bch-gateway";
+    description = "HARDWARIO Gateway (Python Application «bcg»)";
+    platforms = platforms.linux;
+    license = licenses.mit;
+    maintainers = with maintainers; [cynerd];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bcg/default.nix:45]:
```suggestion
    maintainers = with maintainers; [ cynerd ];
```

File: pkgs/development/python-modules/bcf/default.nix [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    repo = "bch-firmware-tool";
+    rev = "v${version}";
+    sha256 = "i28VewTB2XEZSfk0UeCuwB7Z2wz4qPBhzvxJIYkKwJ4=";
+  };
+
+  postPatch = ''
+    sed -ri 's/@@VERSION@@/${version}/g' \
+      bcf/__init__.py setup.py
+  '';
+  doCheck = false;
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bcf/default.nix:43]: This needs a comment

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+      wantedBy = ["multi-user.target"];
+      wants = mkIf config.services.mosquitto.enable ["mosquitto.service"];
+      serviceConfig = {
+        EnvironmentFile = cfg.environmentFiles;
+        ExecStartPre = lib.optional envConfig
+          (pkgs.writeShellScript "pre-start" ''
+            umask 077
+            ${pkgs.envsubst}/bin/envsubst -i "${configFile}" -o "${finalConfig}"
+          '');
+        ExecStart="${cfg.package}/bin/mqtt2influxdb -dc ${finalConfig}";
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:243]:
```suggestion
        ExecStart = "${cfg.package}/bin/mqtt2influxdb -dc ${finalConfig}";
```

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.bigclown-mqtt2influxdb = let
+      envConfig = cfg.environmentFiles != [];
+      finalConfig = if envConfig
+                    then "$RUNTIME_DIRECTORY/mqtt2influxdb.config.yaml"
+                    else configFile;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:231]:
```suggestion
      finalConfig = if envConfig
        then "$RUNTIME_DIRECTORY/mqtt2influxdb.config.yaml"
        else configFile;
```

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+        database = mkOption {
+          type = types.str;
+          description = mdDoc "Name of the InfluxDB database.";
+        };
+        username = mkOption {
+          type = with types; nullOr str;
+          default = null;
+          description = mdDoc "Username for InfluxDB login.";
+        };
+        password = mkOption {
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:202]: This should be passwordFile rather

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+        keyfile = mkOption {
+          type = with types; nullOr path;
+          default = null;
+          description = mdDoc "Key file for MQTT";
+        };
+      };
+      influxdb = {
+        host = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:185]: We can't use sockets?

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+          description = mdDoc "Host where InfluxDB server is running.";
+        };
+        port = mkOption {
+          type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:189]:
```suggestion
          type = types.port;
```

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        };
+        password = mkOption {
+          type = with types; nullOr str;
+          default = null;
+          description = mdDoc ''
+            MQTT password.
+
+            It is highly suggested to use here replacement through
+            environmentFiles as otherwise the password is put world readable to
+            the store.
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:163]: We should only offer that option

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+        '';
+      };
+      mqtt = {
+        host = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
+          description = mdDoc "Host where MQTT server is running.";
+        };
+        port = mkOption {
+          type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:146]:
```suggestion
          type = types.port;
```

File: nixos/modules/services/misc/bcg.nix [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        default = {};
+        description = mdDoc "Rename nodes to different name.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [
+      pkgs.python3Packages.bcg
+      pkgs.python3Packages.bch
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/bcg.nix:151]: Do these really need to be in systemPackages? Can't they be in the service only?

File: nixos/modules/services/misc/bcg.nix [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+        '';
+      };
+      mqtt = {
+        host = mkOption {
+          type = types.str;
+          default = "127.0.0.1";
+          description = mdDoc "Host where MQTT server is running.";
+        };
+        port = mkOption {
+          type = types.int;
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/bcg.nix:75]:
```suggestion
          type = types.port;
```

**Review by SuperSandro2000 — COMMENTED** [review]
Please add pythonImportsCheck's and checkInputs on all python packages.

File: nixos/modules/services/misc/mqtt2influxdb.nix (hunk 9 above)
Comment (Cynerd) [reply nixos/modules/services/misc/mqtt2influxdb.nix:185]: It doesn't seem to be an option if I look at https://github.com/hardwario/bch-mqtt2influxdb/blob/master/mqtt2influxdb/mqtt2influxdb.py#L31 and https://influxdb-python.readthedocs.io/en/latest/api-documentation.html#influxdb.InfluxDBClient. The host is required and needs to be a hostname.

File: nixos/modules/services/misc/mqtt2influxdb.nix (hunk 10 above)
Comment (Cynerd) [reply nixos/modules/services/misc/mqtt2influxdb.nix:189]: 👍

File: nixos/modules/services/misc/mqtt2influxdb.nix (hunk 12 above)
Comment (Cynerd) [reply nixos/modules/services/misc/mqtt2influxdb.nix:146]: 👍

File: nixos/modules/services/misc/bcg.nix (hunk 13 above)
Comment (Cynerd) [reply nixos/modules/services/misc/bcg.nix:151]: They provide control utilities such as listing connected devices or pairing new ones. It makes sense to me to provide them to the user as well.

File: nixos/modules/services/misc/bcg.nix (hunk 14 above)
Comment (Cynerd) [reply nixos/modules/services/misc/bcg.nix:75]: 👍

File: pkgs/development/python-modules/bcg/default.nix [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    owner = "hardwario";
+    repo = "bch-gateway";
+    rev = "v${version}";
+    sha256 = "2Yh5MeIv+BIxjoO9GOPqq7xTAFhyBvnxPy7DeO2FrkI=";
+  };
+
+  postPatch = ''
+    sed -ri 's/@@VERSION@@/${version}/g' \
+      bcg/__init__.py setup.py
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bcg/default.nix:38]: we normally order these three like: src, postPatch, propagatedBuildInputs

File: pkgs/development/python-modules/bcf/default.nix [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    owner = "hardwario";
+    repo = "bch-firmware-tool";
+    rev = "v${version}";
+    sha256 = "i28VewTB2XEZSfk0UeCuwB7Z2wz4qPBhzvxJIYkKwJ4=";
+  };
+
+  postPatch = ''
+    sed -ri 's/@@VERSION@@/${version}/g' \
+      bcf/__init__.py setup.py
+  '';
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/bcf/default.nix:42]: same here

File: nixos/modules/services/misc/mqtt2influxdb.nix [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+      description = "BigClown MQTT to InfluxDB bridge";
+      wantedBy = ["multi-user.target"];
+      wants = mkIf config.services.mosquitto.enable ["mosquitto.service"];
+      serviceConfig = {
+        EnvironmentFile = cfg.environmentFiles;
+        ExecStartPre = lib.optional envConfig
+          (pkgs.writeShellScript "pre-start" ''
+            umask 077
+            ${pkgs.envsubst}/bin/envsubst -i "${configFile}" -o "${finalConfig}"
+          '');
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/mqtt2influxdb.nix:242]: One level higher should be preStart which is already a script and saves us some things here

File: nixos/modules/services/misc/mqtt2influxdb.nix (hunk 8 above)
Comment (SuperSandro2000) [reply nixos/modules/services/misc/mqtt2influxdb.nix:202]: They would need to be substituted before the service is starting.

File: nixos/modules/services/misc/bcg.nix [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+        default = {};
+        description = mdDoc "Rename nodes to different name.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [
+      pkgs.python3Packages.bcg
+      pkgs.python3Packages.bch
```
Comment (SuperSandro2000) [inline nixos/modules/services/misc/bcg.nix:151]:
```suggestion
    environment.systemPackages = with pkgs; [
      python3Packages.bcg
      python3Packages.bch
```
nit

File: pkgs/development/python-modules/bcg/default.nix (hunk 15 above)
Comment (Cynerd) [reply pkgs/development/python-modules/bcg/default.nix]: Ok.

File: pkgs/development/python-modules/bcf/default.nix (hunk 16 above)
Comment (Cynerd) [reply pkgs/development/python-modules/bcf/default.nix]: Ok

File: nixos/modules/services/misc/bcg.nix (hunk 18 above)
Comment (Cynerd) [reply nixos/modules/services/misc/bcg.nix]: Ok

File: nixos/modules/services/misc/mqtt2influxdb.nix (hunk 17 above)
Comment (Cynerd) [reply nixos/modules/services/misc/mqtt2influxdb.nix:248]: Ok

Comment (SuperSandro2000) [reply nixos/modules/services/misc/mqtt2influxdb.nix:248]: You forgot this thread

Comment (Cynerd) [reply nixos/modules/services/misc/mqtt2influxdb.nix:248]: Ups.. sorry. It is addressed now.

**Pull request merged**